crates/uv-auth/src/store.rs

```diff
@@ -0,0 +1,175 @@
+use lazy_static::lazy_static;
+use std::{collections::HashMap, sync::Mutex};
+
+use netrc::Authenticator;
+use tracing::warn;
+use url::Url;
+
+use crate::NetLoc;
+
+lazy_static! {
+    // Store credentials for NetLoc
+    static ref PASSWORDS: Mutex<HashMap<NetLoc, Option<Credential>>> = Mutex::new(HashMap::new());
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub enum Credential {
+    Basic(BasicAuthData),
+    UrlEncoded(UrlAuthData),
+}
+
+impl Credential {
+    pub fn username(&self) -> String {
+        match self {
+            Credential::Basic(auth) => auth.username.clone(),
+            Credential::UrlEncoded(auth) => auth.username.clone(),
+        }
+    }
+    pub fn password(&self) -> Option<String> {
+        match self {
+            Credential::Basic(auth) => auth.password.clone(),
+            Credential::UrlEncoded(auth) => auth.password.clone(),
+        }
+    }
+}
+
+impl From<&Authenticator> for Credential {
+    fn from(auth: &Authenticator) -> Self {
+        Credential::Basic(BasicAuthData {
+            username: auth.login.clone(),
+            password: Some(auth.password.clone()),
+        })
+    }
+}
+
+// Used for URL encoded auth in User info
+// <https://datatracker.ietf.org/doc/html/rfc3986#section-3.2.1>
+#[derive(Clone, Debug, PartialEq)]
+pub struct UrlAuthData {
+    pub username: String,
+    pub password: Option<String>,
+}
+
+impl UrlAuthData {
+    pub fn apply_to_url(&self, mut url: Url) -> Url {
+        url.set_username(&self.username)
+            .unwrap_or_else(|()| warn!("Failed to set username"));
```

Comment:
When can these fail? (These will be effectively silent for the user.)

Reply:
I didn't look into this - it's just copied over from the main (lib.rs) function

Reply:
Okay thanks, that's a perfectly fine answer :)